### PR TITLE
Adds a --lint-head flag to the clang-tidy script

### PR DESCRIPTION
**Pull request by zanderso**
This PR adds a flag to the clang-tidy wrapper script that causes it to lint the files that are changed in the HEAD commit. This will let us avoid linting every file on every commit on prod CI.

In particular, in presubmit, we'll be able to remove the `--lint-all` flag. In post-submit, we'll remove the `lint-all` flag, and use the `--lint-head` flag instead, and only when certain files are changed, like the `DEPS` file, we'll continue passing `--lint-all`.

Related: https://github.com/flutter/flutter/issues/105068

### CODE REVIEW

File: tools/clang_tidy/lib/src/git_repo.dart [hunk 1]
```diff
@@ -64,10 +64,53 @@ class GitRepo {
...
+      defaultWorkingDirectory: root,
+    );
+    await _fetch(processRunner);
+    final ProcessRunnerResult diffTreeResult = await processRunner.runProcess(
+      <String>[
+        'git',
+        'diff-tree',
+        '--no-commit-id',
+        '--name-only',
+        '--diff-filter=ACMRT',
```
Comment (jmagman) [inline tools/clang_tidy/lib/src/git_repo.dart:81]:
```suggestion
        '--diff-filter=ACMRT', # Added, copied, modified, renamed, or type-changed.
```

**Review by jmagman — APPROVED** [review]
LGTM with comment nit

File: tools/clang_tidy/lib/src/git_repo.dart [hunk 2]
```diff
@@ -78,7 +78,7 @@ class GitRepo {
         'diff-tree',
         '--no-commit-id',
         '--name-only',
-        '--diff-filter=ACMRT', # Added, copied, modified, renamed, or type-changed.
```
Comment (jmagman) [inline tools/clang_tidy/lib/src/git_repo.dart]: Doh sorry too many languages to keep track of 🤦‍♀️

Comment (zanderso) [reply tools/clang_tidy/lib/src/git_repo.dart]: No worries! I missed it too...

**Pull request merged**